metabase/metabase
# Prevent repeated coercion when nesting queries

Fixes #47940
Fixes #48721

### Description

This PR does not fix the issue highlighted by the repro steps using join in #47940, but that's not a regression in v50, it fails in v49 too. This is a deeper bug, and fixing it requires recognizing that a reference is from an earlier stage even though it's made by field ID. Such references are bad, IMO and should be eliminated in general.

There are a number of places where `:qp/ignore-coercion` was set, but sometimes they had no effect because `coercion-strategy` was not present and they were completely missing when nesting the query. This PR aims at setting `:qp/ignore-coercion` in all cases when we know the field comes from nesting, even if coercion might not be involved. 

### How to verify

Try the model based repro steps from #47940. To check #48721, use the same table, create a custom column and then use that and the coerced date column in a custom aggregation expression as described in #48721.

### Checklist

- [x] [stress test](https://github.com/metabase/metabase/actions/runs/11892863460/job/33136612909)
